# Our input

A data set containing lots of information about each participent in a speed dating event,such as his  unique subject number,wether he was a male or female, information about each wave such as the number of people he/she met in a wave,station number where met partner,partner's id, wether the participant and the partner were the same race, information about the partner him/her self such as their race,their age,their decision and their rating from 1 to 6...etc.

The dataset also contained information collected from a Survey filled out by students that are interested in participating in order to register for the event Like age,field pf study and each field has its own code in our data set,their Median SAT score for their undergraduate institution and also some questions like How important was it to him/her (on a scale of 1-10) that a person they date be of the same racial/ethnic background or e be of the same religious background, and where were they originally from amd the zip code for the place they grew up in, their goal of paticipating in the event , their career, how happy does he/she expect to be with the people they meet during the speed-dating event .. etc.

It also had informatiom out of a survey Filled out by subjects after each date during the event, the survey required information like their id,their decision and answers for questions about each person they meet like their attributes,how much they like the person..etc.

It has also answers from each participent on Half way through meeting asking about what We want to know what each participent look for in the opposite sex.

Also it has information collected from a Survey filled out the day after participating in the event like how satisfied were they with the people they met,What do they think the opposite sex looks for in a date,How does each participent think he/she measures up...etc.

Last but not least,Subjects filled out 3-4 weeks after they had been sent their matches and the data set included information regarding that such as How many of the matches each paticipent had has he/she contacted to set up a date, Have the participent been on a date with any of his/her matches and how many....etc.





# Our output

Our model is required to predict the outcome of a specific speed dating session based on the profile of two people in order to implement a recommendation system to better match people in speed dating events. we are going to predict the probability that the dating session will lead to a successful match.

# What data mining function is required?

binary classification using pipelines,grid search,Random Search and Bayesian Search.

# What could be the challenges?

Developing a successful solution to our problem , complex data,datasets can include complex data elements ,another thing is that we have to make sure that our algorithm must be efficient and scalable to extract information from the big data and we should have enough knowledge and experience in order to use them if we needed to improve our algorithms.

# What is the impact?

Our model predictions is going to implement a recommendation system to better match people in speed dating events which will make it easier when it comes to dating 

# What is an ideal solution?

 An ideal solution in my opinion will be measured in terms of metrics and performances

# What is the experimental protocol used and how was it carried out? What preprocessing steps are used?

when it comes to the preprocessing steps I made a pipeline performing the following on the training set and testing set :

SimpleImputer(),Imputation transformer for completing missing values.

one hot encoder to transform categorical data to numerical data

stadnard scalar to standard scale my data 

The experimental protocol I used in my code is cross validation and validation set

# Why a simple linear regression model (without any activation function) is not good for classification task, compared to Perceptron/Logistic regression?

Continuous values are predicted and dealt with by linear regression, whereas discrete values are predicted and dealt with by classification. The second issue is that when new data points are added, the threshold value shifts.

While a Categorical value, such as 0 or 1, Yes or No, is the result of Logistic Regression/perceptron.


# What's a decision tree and how it is different to a logistic regression model?

A decision tree is a tool with a tree-like structure that predicts likely outcomes, resource costs, utility costs, and potential implications. Decision trees are a mechanism to provide conditional control assertions in algorithms. They have branches that indicate decision-making steps that could result in a positive outcome.

Decision Trees divide the space into smaller and smaller sections, whereas Logistic Regression fits a single line to precisely divide the space into two. These lines would, of course, generalise to planes and hyperplanes for higher-dimensional data.

# What's the difference between grid search and random search?

GridSearchCV differs from RandomizedSearchCV in that in Grid Search, we test every combination of a preset list of hyper-parameter values and chose the best one based on the cross-validation score. Random search tries a variety of different values at random (we have to define the number iterations).

# What's the difference between bayesian search and random search?

Because they select hyperparameters in an informed manner, Bayesian optimization methods are efficient. Bayesian approaches can locate the best hyperparameters in less time (fewer iterations) than grid search and random search because they prioritise hyperparameters that appear more promising from previous findings.

In [1]:
import pandas as pd #I imported pandas cause Using it
#make it way easier when it comes to  processing, analysis and manipulation of data
from sklearn.feature_selection import SelectKBest, chi2

In [2]:
df = pd.read_csv('train.csv') #I imported my training data set
# Set column "id" as index column
df.set_index("id", inplace = True)
print(df) #printing my training data set

      gender  idg  condtn  wave  round  position  positin1  order  partner  \
id                                                                           
2583       0    3       2    14     18         2       2.0     14       12   
6830       1   14       1     3     10         2       NaN      8        8   
4840       1   14       1    13     10         8       8.0     10       10   
5508       1   38       2     9     20        18      13.0      6        7   
4828       1   24       2    14     20         6       6.0     20       17   
...      ...  ...     ...   ...    ...       ...       ...    ...      ...   
3390       0    1       2     9     20         2       2.0     18        1   
4130       1   24       2     9     20        19      15.0      5        6   
1178       0   13       2    11     21         5       5.0      3       18   
5016       1   10       2     7     16         6      14.0      9       10   
8149       0    7       2    21     22         7       7.0      

In [3]:
df2 = pd.read_csv('test.csv') #I imported my testing data set
# Set column "id" as index column
df2.set_index("id", inplace = True)
print(df2) #printing my testing data set

      gender  idg  condtn  wave  round  position  positin1  order  partner  \
id                                                                           
934        0    5       2     2     16         3       NaN     13       13   
6539       0   33       2    14     18         6       6.0      4        8   
6757       1    6       2     9     20        10      16.0     15       19   
2275       1   26       2     2     19        15       NaN      8       10   
1052       0   29       2     7     16         7       7.0     10        5   
...      ...  ...     ...   ...    ...       ...       ...    ...      ...   
7982       0   23       2    15     19        18      18.0     14       11   
7299       0    5       1    13      9         4       4.0      4        8   
1818       1   26       2     2     19         3       NaN     15        3   
937        0   19       2     9     20        11      11.0      9        2   
6691       1   38       2    21     22        22       7.0     1

In [4]:
print(df.shape)#shape of training set
print(df2.shape)#shape of testing set

(5909, 191)
(2469, 190)


In [5]:
# check for missing data
df.isnull().sum().sort_values(ascending=False)

num_in_3    5449
numdat_3    4849
expnum      4627
amb7_2      4519
sinc7_2     4519
            ... 
order          0
partner        0
match          0
samerace       0
gender         0
Length: 191, dtype: int64

In [6]:
#same for testing set
df2.isnull().sum().sort_values(ascending=False)

num_in_3    2261
numdat_3    2033
expnum      1951
sinc7_2     1904
amb7_2      1904
            ... 
position       0
round          0
wave           0
condtn         0
gender         0
Length: 190, dtype: int64

In [7]:
#we have a lot of nulls

In [8]:
# checking types of columns in trainig set :
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5909 entries, 2583 to 8149
Columns: 191 entries, gender to amb5_3
dtypes: float64(173), int64(10), object(8)
memory usage: 8.7+ MB


In [9]:
# checking types of columns in testing set :
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2469 entries, 934 to 6691
Columns: 190 entries, gender to amb5_3
dtypes: float64(173), int64(9), object(8)
memory usage: 3.6+ MB


In [10]:
#seems like I have 8 object columns, I'm going to get the names of those object columns,
catCols = [col for col in df.columns if df[col].dtype=="O"]
#display their names
catCols

['field',
 'undergra',
 'mn_sat',
 'tuition',
 'from',
 'zipcode',
 'income',
 'career']

In [11]:
df['field'] #display field column

id
2583    Ed.D. in higher education policy at TC
6830                               Engineering
4840                            Urban Planning
5508                     International Affairs
4828                                  Business
                         ...                  
3390                       Clinical Psychology
4130                                       MBA
1178                      MA Science Education
5016                              Biochemistry
8149                        MFA Acting Program
Name: field, Length: 5909, dtype: object

In [12]:
#for those object columns,I'll use categorical encoding to change their types from object to category in both training and
#testing data
df['field'] = df['field'].astype("category")
df2['field'] = df2['field'].astype("category")
df2['undergra'] = df2['undergra'].astype("category")
df['undergra'] = df['undergra'].astype("category")
df['mn_sat'] = df['mn_sat'].astype("category")
df2['mn_sat'] = df2['mn_sat'].astype("category")
df['tuition'] = df['tuition'].astype("category")
df2['tuition'] = df2['tuition'].astype("category")
df['from'] = df['from'].astype("category")
df2['from'] = df2['from'].astype("category")
df['zipcode'] = df['zipcode'].astype("category")
df2['zipcode'] = df2['zipcode'].astype("category")
df['income'] = df['income'].astype("category")
df2['income'] = df2['income'].astype("category")
df['career'] = df['career'].astype("category")
df2['career'] = df2['career'].astype("category")


In [13]:
print(df) #printing my training set after categorical enconding for its 8 object columns

      gender  idg  condtn  wave  round  position  positin1  order  partner  \
id                                                                           
2583       0    3       2    14     18         2       2.0     14       12   
6830       1   14       1     3     10         2       NaN      8        8   
4840       1   14       1    13     10         8       8.0     10       10   
5508       1   38       2     9     20        18      13.0      6        7   
4828       1   24       2    14     20         6       6.0     20       17   
...      ...  ...     ...   ...    ...       ...       ...    ...      ...   
3390       0    1       2     9     20         2       2.0     18        1   
4130       1   24       2     9     20        19      15.0      5        6   
1178       0   13       2    11     21         5       5.0      3       18   
5016       1   10       2     7     16         6      14.0      9       10   
8149       0    7       2    21     22         7       7.0      

In [14]:
print(df2) #printing my testing set after categorical enconding for its 8 object columns

      gender  idg  condtn  wave  round  position  positin1  order  partner  \
id                                                                           
934        0    5       2     2     16         3       NaN     13       13   
6539       0   33       2    14     18         6       6.0      4        8   
6757       1    6       2     9     20        10      16.0     15       19   
2275       1   26       2     2     19        15       NaN      8       10   
1052       0   29       2     7     16         7       7.0     10        5   
...      ...  ...     ...   ...    ...       ...       ...    ...      ...   
7982       0   23       2    15     19        18      18.0     14       11   
7299       0    5       1    13      9         4       4.0      4        8   
1818       1   26       2     2     19         3       NaN     15        3   
937        0   19       2     9     20        11      11.0      9        2   
6691       1   38       2    21     22        22       7.0     1

In [15]:
# checking types again:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5909 entries, 2583 to 8149
Columns: 191 entries, gender to amb5_3
dtypes: category(8), float64(173), int64(10)
memory usage: 8.5 MB


In [16]:
#Importing numpy library and Splitting the data into features and target variables
import numpy as np
Xtr=df.drop('match', axis=1)#here I started splitting my training set (df) into Xtr (input)
ytr=df['match'] #and ytr the output (my target) which is the matching
Xts=df2 #my xtest which is equal to the testing dataframe
Xtr #displaying my x train

gender  idg  condtn  wave  round  position  positin1  order  partner  \
id                                                                           
2583       0    3       2    14     18         2       2.0     14       12   
6830       1   14       1     3     10         2       NaN      8        8   
4840       1   14       1    13     10         8       8.0     10       10   
5508       1   38       2     9     20        18      13.0      6        7   
4828       1   24       2    14     20         6       6.0     20       17   
...      ...  ...     ...   ...    ...       ...       ...    ...      ...   
3390       0    1       2     9     20         2       2.0     18        1   
4130       1   24       2     9     20        19      15.0      5        6   
1178       0   13       2    11     21         5       5.0      3       18   
5016       1   10       2     7     16         6      14.0      9       10   
8149       0    7       2    21     22         7       7.0      2       12   

        pid  ...  attr3_3  sinc3_3  intel3_3  fun3_3  amb3_3  attr5_3  \
id           ...                                                        
2583  372.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
6830   63.0  ...      6.0      8.0       8.0     7.0     8.0      NaN   
4840  331.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
5508  200.0  ...      8.0      9.0       8.0     8.0     6.0      NaN   
4828  357.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
...     ...  ...      ...      ...       ...     ...     ...      ...   
3390  214.0  ...     12.0     12.0      12.0     9.0    12.0      NaN   
4130  199.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
1178  290.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
5016  151.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
8149  542.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   

      sinc5_3  intel5_3  fun5_3  amb5_3  
id                                       
2583      NaN       NaN     NaN     NaN  
6830      NaN       NaN     NaN     NaN  
4840      NaN       NaN     NaN     NaN  
5508      NaN       NaN     NaN     NaN  
4828      NaN       NaN     NaN     NaN  
...       ...       ...     ...     ...  
3390      NaN       NaN     NaN     NaN  
4130      NaN       NaN     NaN     NaN  
1178      NaN       NaN     NaN     NaN  
5016      NaN       NaN     NaN     NaN  
8149      NaN       NaN     NaN     NaN  

[5909 rows x 190 columns]

In [17]:
Xts #displaying x test

gender  idg  condtn  wave  round  position  positin1  order  partner  \
id                                                                           
934        0    5       2     2     16         3       NaN     13       13   
6539       0   33       2    14     18         6       6.0      4        8   
6757       1    6       2     9     20        10      16.0     15       19   
2275       1   26       2     2     19        15       NaN      8       10   
1052       0   29       2     7     16         7       7.0     10        5   
...      ...  ...     ...   ...    ...       ...       ...    ...      ...   
7982       0   23       2    15     19        18      18.0     14       11   
7299       0    5       1    13      9         4       4.0      4        8   
1818       1   26       2     2     19         3       NaN     15        3   
937        0   19       2     9     20        11      11.0      9        2   
6691       1   38       2    21     22        22       7.0     16        5   

        pid  ...  attr3_3  sinc3_3  intel3_3  fun3_3  amb3_3  attr5_3  \
id           ...                                                        
934    52.0  ...      5.0      7.0       8.0     6.0     8.0      NaN   
6539  368.0  ...      6.0      8.0       7.0     7.0     8.0      6.0   
6757  212.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
2275   30.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
1052  162.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
...     ...  ...      ...      ...       ...     ...     ...      ...   
7982  407.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
7299  339.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
1818   23.0  ...      NaN      NaN       NaN     NaN     NaN      NaN   
937   215.0  ...      9.0      7.0      12.0    12.0     9.0      NaN   
6691  513.0  ...      7.0      9.0       8.0     7.0     8.0      5.0   

      sinc5_3  intel5_3  fun5_3  amb5_3  
id                                       
934       NaN       NaN     NaN     NaN  
6539      7.0       6.0     5.0     5.0  
6757      NaN       NaN     NaN     NaN  
2275      NaN       NaN     NaN     NaN  
1052      NaN       NaN     NaN     NaN  
...       ...       ...     ...     ...  
7982      NaN       NaN     NaN     NaN  
7299      NaN       NaN     NaN     NaN  
1818      NaN       NaN     NaN     NaN  
937       NaN       NaN     NaN     NaN  
6691      8.0       8.0     6.0     8.0  

[2469 rows x 190 columns]

In [18]:
ytr #displaying my y train

id
2583    0
6830    0
4840    0
5508    0
4828    0
       ..
3390    0
4130    0
1178    0
5016    1
8149    0
Name: match, Length: 5909, dtype: int64

In [19]:
#now the pipeline
# we extract numeric features and categorical features names

# numeric features can be selected by: (based on the df2.info() output )
features_numeric = list(Xtr.select_dtypes(include=['float64', 'int64']))

# categorical features can be selected by: (based on the df2.info() output )
features_categorical = list(Xtr.select_dtypes(include=['category']))

print('numeric features:', features_numeric) #printing numeric features
print('categorical features:', features_categorical) #printing categorical features

numeric features: ['gender', 'idg', 'condtn', 'wave', 'round', 'position', 'positin1', 'order', 'partner', 'pid', 'int_corr', 'samerace', 'age_o', 'race_o', 'pf_o_att', 'pf_o_sin', 'pf_o_int', 'pf_o_fun', 'pf_o_amb', 'pf_o_sha', 'attr_o', 'sinc_o', 'intel_o', 'fun_o', 'amb_o', 'shar_o', 'like_o', 'prob_o', 'met_o', 'age', 'field_cd', 'race', 'imprace', 'imprelig', 'goal', 'date', 'go_out', 'career_c', 'sports', 'tvsports', 'exercise', 'dining', 'museums', 'art', 'hiking', 'gaming', 'clubbing', 'reading', 'tv', 'theater', 'movies', 'concerts', 'music', 'shopping', 'yoga', 'exphappy', 'expnum', 'attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1', 'shar1_1', 'attr4_1', 'sinc4_1', 'intel4_1', 'fun4_1', 'amb4_1', 'shar4_1', 'attr2_1', 'sinc2_1', 'intel2_1', 'fun2_1', 'amb2_1', 'shar2_1', 'attr3_1', 'sinc3_1', 'fun3_1', 'intel3_1', 'amb3_1', 'attr5_1', 'sinc5_1', 'intel5_1', 'fun5_1', 'amb5_1', 'attr', 'sinc', 'intel', 'fun', 'amb', 'shar', 'like', 'prob', 'met', 'match_es', 'attr1_s', 'sin

In [20]:
#importing important libraries
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.datasets import fetch_openml
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV

np.random.seed(0) #every time you call the numpy's other random function, the result will be the same

# define a pipe line for numeric feature preprocessing
# we gave them a name so we can set their hyperparameters
transformer_numeric = Pipeline(
    steps=[
        ('imputer', SimpleImputer()),#Imputation transformer for completing missing values.
        ('scaler', StandardScaler())]#Standardize features

)

# define a pipe line for categorical feature preprocessing
# we gave them a name so we can set their hyperparameters
transformer_categorical = Pipeline(
    steps=[
        ('imputer', SimpleImputer(strategy='constant')), #here I set the strategy for completting missing values  = constant
        ('onehot', OneHotEncoder(handle_unknown='ignore')) #Encode categorical features as a one-hot numeric
        #When handle_unknown parameter is set to ‘ignore’ and an unknown category is encountered during transform,
        #the resulting one-hot encoded columns for this feature will be all zeros.

    ]
)
# define the preprocessor 
# we gave them a name so we can set their hyperparameters
# we also specify what are the categorical 
preprocessor = ColumnTransformer(
    transformers=[
        ('num', transformer_numeric, features_numeric),
        ('cat', transformer_categorical, features_categorical)
    ]
)

# combine the preprocessor with the model as a full tunable pipeline
# we gave them a name so we can set their hyperparameters
full_pipline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('my_classifier', 
           RandomForestClassifier(), #I want to use random forest classifier
        )
    ]
)
full_pipline 

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['gender', 'idg', 'condtn',
                                                   'wave', 'round', 'position',
                                                   'positin1', 'order',
                                                   'partner', 'pid', 'int_corr',
                                                   'samerace', 'age_o',
                                                   'race_o', 'pf_o_att',
                                                   'pf_o_sin', 'pf_o_int',
                                                   

In [21]:
# The pipeline object can be used like any sk-learn model
full_pipline = full_pipline.fit(Xtr, ytr) #fitting the data in the pipe
full_pipline.predict(Xts) #predicting

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [22]:
# Grid Search with Cross-validation
# here we specify the search space
param_grid = {
    'preprocessor__num__imputer__strategy': ['median'], #I set it to median
    'my_classifier__n_estimators': [50, 100, 150],  # ranges of n_estimators which are the number of trees to be used in the forest.
    # I set my  n_estimators ranges to [50, 100, 150]
    'my_classifier__max_depth':[30, 60, 90]  
    # I set my  max depth ranges to [30, 60, 90] which are The number of splits that each decision tree is allowed to make.
}

# two-fold cross-validation
# n_jobs means cucurrent number of jobs is 2
grid_search = GridSearchCV(
    full_pipline, param_grid, cv=2, verbose=1 # showing more 'wordy' information
    , n_jobs=2, 
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

grid_search.fit(Xtr, ytr) #fitting Xtr and ytr

print('best score {}'.format(grid_search.best_score_)) #printing best score
print('best parameters {}'.format(grid_search.best_params_))#printing best parameters


Fitting 2 folds for each of 9 candidates, totalling 18 fits
best score 0.8243617887449464
best parameters {'my_classifier__max_depth': 30, 'my_classifier__n_estimators': 150, 'preprocessor__num__imputer__strategy': 'median'}


In [23]:
# the resulting model is the same pipeline but with the best hyperparameters
# trained on the full training set. we can use it directly
ypred=grid_search.predict(Xts) #predicting ypred

In [24]:
submission = pd.DataFrame() # creating dataframe for the submission
submission['id'] = Xts.index #using the column index of x test values to fill submission['id'] column
submission['match'] = grid_search.predict_proba(Xts)[:,1] #predicting the probabilities and filling submission['match'] column with their values
submission.to_csv('sample_submission_walkthrough.csv', index=False)#generating the submission file


# Trial one using RandomForest with Grid Search with Cross-validation 


Cross-validation is a method for robustly estimating test-set performance (generalization) of a model. Grid-searching is the process of scanning the data to configure optimal parameters for a given model.



# Expectation

I expect to get the optimal hyperparamters that will give me the best performance and highest accuracy using random forest model with grid search and cross validation

# Observation 

After running the code,the best hyperparameters that was defined by the grid search using cross validation were : 
    
max_depth: 30
    
n_estimators: 150
    
with using imputer__strategy of: 'median'
    
The model gave me a best score of 0.8243617887449464

Score on kaggle 0.84021

# Plan

I'm going to set different ranges for the hyperparameters in the grid search with cross validation


n_estimators': [90, 180, 270 ]
    
    
max_depth':[30, 100, 170]
    
    
and I changed my imputer__strategy' to 'most_frequent'


and I changed number of folds to cv=4

In [25]:
# after changing hyperparameters ranges/values
param_grid = {
    'preprocessor__num__imputer__strategy': ['most_frequent'], # this is the strategy of completing missing data it will be filled with most frequent  
    'my_classifier__n_estimators': [90, 180, 270], #I changed my ranges of the number of trees to be used in the forest into: [90,180,270]
    'my_classifier__max_depth':[30, 100, 170]
    #  I set my  max depth ranges to [30, 100, 170] which are The number of splits that each decision tree is allowed to make.
    
}

# cv=2 means two-fold cross-validation
# n_jobs means the cucurrent number of jobs
grid_search = GridSearchCV(
    full_pipline, param_grid, cv=4, verbose=1,# showing more 'wordy' information
    n_jobs=2, #n_jobs means cucurrent number of jobs is 2
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

grid_search.fit(Xtr, ytr) #fitting Xtr and ytr

print('best score {}'.format(grid_search.best_score_))
print('best parameters {}'.format(grid_search.best_params_))

Fitting 4 folds for each of 9 candidates, totalling 36 fits
best score 0.8355541226745274
best parameters {'my_classifier__max_depth': 100, 'my_classifier__n_estimators': 270, 'preprocessor__num__imputer__strategy': 'most_frequent'}


In [26]:
# the resulting model is the same pipeline with the best hyperparameters
# trained on the full training set. we can use it directly
ypred=grid_search.predict(Xts) #predicting results

In [27]:
submission = pd.DataFrame()# creating dataframe for the submission
submission['id'] = Xts.index#using the column index of x test values to fill submission['id'] column
submission['match'] = grid_search.predict_proba(Xts)[:,1]#predicting the probabilities and filling submission['match'] column with their values
submission.to_csv('sample_submission_walkthrough.csv', index=False)#generating the submission file

# Trial 2 

# Expectation

I expected to get a higher accuracy after changing the ranges/values of my model's hyperparameters

# Observation 

After running the code,the best hyperparameters that was defined by the grid search using cross validation were : 
    
max_depth: 100
    
n_estimators: 270
    
with using imputer__strategy of: 'most_frequent'
    
The model gave me a best score of 0.8355541226745274

Score on kaggle 0.84440



# Plan

I will still use randomforest classifier, but this time with random search with validation set and see how my accuracy gets affected

In [29]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV #i imported the train_test_split to Further split
#the original training set to a train and a validation set
#I also imported the RandomizedSearchCv to perrform the random search
from sklearn.model_selection import PredefinedSplit #PredefinedSplit Provides train/test indices to split data into train/test sets
#using a predefined scheme specified by user with the test_fold parameter.
param_grid = {
    'preprocessor__num__imputer__strategy': ['constant'], #this is the strategy of completing missing data, it will be filled with constant 
    'my_classifier__n_estimators': [20, 30, 40],  #I changed my ranges to the number of trees to be used in the forest into: [20,30,40]
    'my_classifier__max_depth':[100,200,300]#I set my  max depth ranges to [100, 200, 300] which are The number of splits that each decision tree is allowed to make.      
}

# Further split the original training set to a train and a validation set
X_train2, X_val, y_train2, y_val = train_test_split(
    Xtr, ytr, train_size = 0.8, stratify = ytr, random_state = 2022)#the random_state parameter is used for initializing the internal random number generator
#The stratify option instructs sklearn to divide the dataset into a test and training set with the ratio of class labels in the variable supplied.

# Create a list where train data indices are -1 and validation data indices are 0
# X_train2 (new training set), X_train
split_index = [-1 if x in X_train2.index else 0 for x in Xtr.index]

# Use the list to create PredefinedSplit
pds = PredefinedSplit(test_fold = split_index)
grid_search = RandomizedSearchCV( #grid_search
    full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2,
    # number of random trials
    n_iter=10,
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

grid_search.fit(Xtr, ytr) #fitting my data into grid_search results

print('best score {}'.format(grid_search.best_score_))
print('best parameters {}'.format(grid_search.best_params_))

C:\Users\Debi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:285: UserWarning: The total space of parameters 9 is smaller than n_iter=10. Running 9 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 1 folds for each of 9 candidates, totalling 9 fits
best score 0.8066539377514987
best parameters {'preprocessor__num__imputer__strategy': 'constant', 'my_classifier__n_estimators': 40, 'my_classifier__max_depth': 100}


In [30]:
# the resulting model is the same pipeline with the best hyperparameters
# trained on the full training set. we can use it directly
ypred=grid_search.predict(Xts)

In [31]:
submission = pd.DataFrame()# creating dataframe for the submission
submission['id'] = Xts.index#using the column index of x test values to fill submission['id'] column
submission['match'] = grid_search.predict_proba(Xts)[:,1]#predicting the probabilities and filling submission['match'] column with their values
submission.to_csv('sample_submission_walkthrough.csv', index=False)#generating the submission file

# Trial 3 with using randomsearch with validation set

random search CV give us local optimal (may be good enough and even more generalizable)

# Expectation

I expected to get a higher score after using randomsearch with validation set

# Observation

After running the code,the best hyperparameters that was defined by randomsearch with validation set were : 

n_estimators: 40

max_depth: 100

with using imputer__strategy: 'constant'

The model gave me a best score of 0.8066539377514987

Score on kaggle 0.82700



# Plan

I will still use random search with  validation set

bur I will change 'imputer__strategy': ['median'],

n_estimators': [100, 120, 140], 

max_depth':[5,10,15] ,

n_iter=6

In [36]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.model_selection import PredefinedSplit
param_grid = {
    'preprocessor__num__imputer__strategy': ['median'],
#this is the strategy of completing missing data, it will be filled with the median 
    'my_classifier__n_estimators': [100, 120, 140],  
      #I changed my ranges to the number of trees to be used in the forest into: [100,120,140]
    'my_classifier__max_depth':[5,10,15] #I set my  max depth ranges to [5, 10, 15] which are The number of splits that each decision tree is allowed to make.        
}


grid_search = RandomizedSearchCV(
    full_pipline, param_grid, cv=pds, verbose=1,# showing more 'wordy' information
    n_jobs=2, # n_jobs means the cucurrent number of jobs
    # number of random trials
    n_iter=6,
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

grid_search.fit(Xtr, ytr)

print('best score {}'.format(grid_search.best_score_)) #printing best scores
print('best parameters {}'.format(grid_search.best_params_)) #print best parameters

Fitting 1 folds for each of 6 candidates, totalling 6 fits
best score 0.8298636774246531
best parameters {'preprocessor__num__imputer__strategy': 'median', 'my_classifier__n_estimators': 140, 'my_classifier__max_depth': 5}


In [37]:
# the resulting model is the same pipeline with the best hyperparameters
# trained on the full training set. we can use it directly
ypred=grid_search.predict(Xts)

In [38]:
submission = pd.DataFrame()# creating dataframe for the submission
submission['id'] = Xts.index#using the column index of x test values to fill submission['id'] column
submission['match'] = grid_search.predict_proba(Xts)[:,1]#predicting the probabilities and filling submission['match'] column with their values
submission.to_csv('sample_submission_walkthrough.csv', index=False)#generating the submission file

# Trial 4

# Expectation

I expected to get a higher score after changing the ranges/values of my model's hyperparameters

# Observation

After running the code,the best hyperparameters that was defined by randomsearch with validation set were : 

n_estimators: 140

max_depth: 5

with using imputer__strategy: 'median'

The model gave me a best score of 0.8298636774246531

Score on kaggle 0.82827

my best score and score on kaggle both increased

# plan

I will implement support vector machine model using bayesian search to get the best hyperparameters

In [39]:
# Let's try this with SVM model
from skopt import BayesSearchCV #for Bayesian optimization over hyper parameters.
from skopt.space import Real, Categorical, Integer
from sklearn.svm import SVC #to implement svc model

SVC_pipline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('my_svc', SVC(class_weight='balanced',probability=True)) #setting probability to true to be able to use probna() later when submission
        #class_weight='balanced' assigns the class weights inversely proportional to their respective frequencies
    ]
)


# define ranges for bayes search
bayes_search = BayesSearchCV(
    SVC_pipline,
    {
        'my_svc__C': Real(1e-8, 1e+8, prior='log-uniform'), #ranges for c which is the Penalty parameter of the error term.
        'my_svc__gamma': Real(1e-9, 1e+1, prior='log-uniform'), #ranges for gamma, the gamma parameter defines how far the influence of a single training example reaches
        'my_svc__degree': Integer(1,8), #degree ranges,the degree of the polynomial kernel function 
        'my_svc__kernel': Categorical(['poly','linear', 'rbf']), #kernel ranges,A kernel is a function used in SVM for helping to solve problems
     
    },
    # number of trials 
    n_iter=6,
    random_state=0,#the random_state parameter is used for initializing the internal random number generator
    verbose=1,# showing more 'wordy' information
    cv=pds,
)

bayes_search.fit(Xtr, ytr)

print('best score {}'.format(bayes_search.best_score_))
print('best parameters {}'.format(bayes_search.best_params_))

Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
best score 0.8324873096446701
best score OrderedDict([('my_svc__C', 2.2841098739835276e-06), ('my_svc__degree', 6), ('my_svc__gamma', 5.005418606010382e-05), ('my_svc__kernel', 'poly')])


In [40]:
# the resulting model is the same pipeline with the best hyperparameters
# trained on the full training set. we can use it directly
ypred=bayes_search.predict(Xts) #predicting results

In [42]:
submission = pd.DataFrame() # creating dataframe for the submission
submission['id'] = Xts.index #using the column index of x test values to fill submission['id'] column
submission['match'] = bayes_search.predict_proba(Xts)[:,1] #predicting the probabilities and filling submission['match'] column with their values
submission.to_csv('sample_submission_walkthrough.csv', index=False)#generating the submission file

# Trial 5

I used support vector machine using bayesian search to get the best hyperparameters

Support vector machines are supervised learning models with associated learning algorithms that analyze data for classification and regression analysis. 

Bayesian optimization, a model-based method for finding the minimum of a function, has recently been applied to machine learning hyperparameter tuning, with results suggesting this approach can achieve better performance on the test set while requiring fewer iterations than random search.

# Expectation


I expected to get a higher score after using support vector machines with bayesian search

# Observation

After running the code,the best hyperparameters that was defined by bayesian search were : 
    
C= 2.2841098739835276e-06

degree = 6

gamma=  5.005418606010382e-05

kernel = 'poly'


The model gave me a best score of 0.8324873096446701

Score on kaggle 0.50271



# plan

I will change ranges of hyperparameters to 

'C': Real(2e-5, 1e+10, prior='log-uniform')

'gamma': Real(3e-6, 2e+4, prior='log-uniform')

and i will set 

'kernel': ['poly']

In [43]:
SVC_pipline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('my_svc', SVC(class_weight='balanced',probability=True)) #setting probability to true to be able to use probna() later when submission
        #class_weight='balanced' assigns the class weights inversely proportional to their respective frequencies
    ]
)


# define ranges for bayes search
bayes_search = BayesSearchCV(
    SVC_pipline,
    {
        'my_svc__C': Real(2e-5, 1e+10, prior='log-uniform'), #ranges for c which is the Penalty parameter of the error term.
        'my_svc__gamma': Real(3e-6, 2e+4, prior='log-uniform'), #ranges for gamma, the gamma parameter defines how far the influence of a single training example reaches
        'my_svc__degree': Integer(1,8), #degree ranges, the degree of the polynomial kernel function 
        'my_svc__kernel': Categorical([ 'poly']), #kernel ranges,A kernel is a function used in SVM for helping to solve problems
     
    },
    # number of trials 
    n_iter=6,
    random_state=0,#the random_state parameter is used for initializing the internal random number generator
    verbose=1,# showing more 'wordy' information
    cv=pds,
)

bayes_search.fit(Xtr, ytr)

print('best score {}'.format(bayes_search.best_score_))
print('best parameters {}'.format(bayes_search.best_params_))

Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
best score 0.8282571912013537
best parameters OrderedDict([('my_svc__C', 0.052329516515125955), ('my_svc__degree', 3), ('my_svc__gamma', 42.72743322299447), ('my_svc__kernel', 'poly')])


In [44]:
# the resulting model is the same pipeline with the best hyperparameters
# trained on the full training set. we can use it directly
ypred=bayes_search.predict(Xts) #predicting results

In [45]:
submission = pd.DataFrame() # creating dataframe for the submission
submission['id'] = Xts.index #using the column index of x test values to fill submission['id'] column
submission['match'] = bayes_search.predict_proba(Xts)[:,1] #predicting the probabilities and filling submission['match'] column with their values
submission.to_csv('sample_submission_walkthrough.csv', index=False)#generating the submission file

# Trial 6

I changed ranges of hyperparameters to 

'C': Real(2e-5, 1e+10, prior='log-uniform')

'gamma': Real(3e-6, 2e+4, prior='log-uniform')

and i will set 

'kernel': ['poly']

# Expectation


I expected to get a higher score after changing the ranges of my support vector machine model

# Observation

After running the code,the best hyperparameters that was defined by bayesian search were : 
    
C= 0.052329516515125955

degree = 3

gamma=  42.72743322299447

with using kernel = 'poly'


The model gave me a best score of 0.8282571912013537

Score on kaggle 0.81313

